src/NosCore.Controllers/ExchangePacketController.cs

```diff
@@ -0,0 +1,295 @@
+            {
+                case RequestExchangeType.Requested:
+                    _exchangeService.RequestExchange(Session, target.Session);
+                    return;
+
+                case RequestExchangeType.List:
+                    if (!_exchangeService.OpenExchange(Session.Character.VisualId, target.CharacterId))
+                    {
+                        return;
+                    }
+                    Session.SendPacket(new ServerExcListPacket
+                    {
+                        VisualType = VisualType.Player,
+                        VisualId = target.VisualId,
+                        Gold = null
+                    });
+
+                    target.SendPacket(new ServerExcListPacket
+                    {
+                        VisualType = VisualType.Player,
+                        VisualId = Session.Character.VisualId,
+                        Gold = null
+                    });
+                    return;
+
+                case RequestExchangeType.Declined:
+                    Session.SendPacket(Session.Character.GenerateSay(Language.Instance.GetMessageFromKey(LanguageKey.EXCHANGE_REFUSED, Session.Account.Language), SayColorType.Yellow));
+                    target?.SendPacket(target.GenerateSay(target.GetMessageFromKey(LanguageKey.EXCHANGE_REFUSED), SayColorType.Yellow));
+                    return;
+
+                case RequestExchangeType.Confirmed:
+                    var targetId = _exchangeService.GetTargetId(Session.Character.CharacterId);
+
+                    if (!targetId.HasValue)
+                    {
+                        _logger.Error(LogLanguage.Instance.GetMessageFromKey(LanguageKey.INVALID_EXCHANGE));
+                        return;
+                    }
+                    
+                    target = Broadcaster.Instance.GetCharacter(s => s.VisualId == targetId.Value && s.MapInstance == Session.Character.MapInstance) as Character;
+
+                    if (target == null)
+                    {
+                        _logger.Error(LogLanguage.Instance.GetMessageFromKey(LanguageKey.CANT_FIND_CHARACTER));
+                        return;
+                    }
+
+                    var success = true;
+                    _exchangeService.ConfirmExchange(Session.Character.VisualId);
+
+                    if (!_exchangeService.IsExchangeConfirmed(Session.Character.VisualId) || !_exchangeService.IsExchangeConfirmed(target.VisualId))
+                    {
+                        Session.SendPacket(new InfoPacket { Message = Language.Instance.GetMessageFromKey(LanguageKey.IN_WAITING_FOR, Session.Account.Language) });
+                        return;
+                    }
+
+                    var exchangeInfo = _exchangeService.GetData(Session.Character.CharacterId);
+                    var targetInfo = _exchangeService.GetData(target.CharacterId);
+
+                    if (exchangeInfo.Gold + target.Gold > _worldConfiguration.MaxGoldAmount)
```

Comment:
move all this till if(success) to the service in a 
methode ValidateExchangeData that return tuple<success, packet>